lib/WUI/espif.c

```diff
@@ -0,0 +1,521 @@
+                log_error(ESPIF, "error!");
+                return ERR_MEM;
+            }
+        }
+    } else {
+        log_error(ESPIF, "Already initialized");
+    }
+
+    espif_set_operating_mode(ESPIF_RUNNING_MODE);
+    initialized = 1;
+
+    return ERR_OK;
+}
+
+static err_t low_level_deinit() {
+    if (uartBufferMbox_id != NULL) {
+        osMessageQId tmp = uartBufferMbox_id;
+        uartBufferMbox_id = NULL;
+        osMessageDelete(tmp);
+    }
+    if (UartBufferThread_id != NULL) {
+        osThreadId tmp = UartBufferThread_id;
+        UartBufferThread_id = NULL;
+        osThreadTerminate(tmp);
+    }
+    initialized = 0;
+    return ERR_OK;
+}
+
+err_t espif_flash_initialize() {
+    err_t err = low_level_deinit();
+    if (err != ERR_OK) {
+        return err;
+    }
+    espif_set_operating_mode(ESPIF_FLASHING_MODE);
+    espif_reconfigure_uart(115200);
+    loader_stm32_config_t loader_config = {
+        .huart = &huart6,
+        .port_io0 = GPIOE,
+        .pin_num_io0 = GPIO_PIN_6,
+        .port_rst = GPIOC,
+        .pin_num_rst = GPIO_PIN_13,
+    };
+    loader_port_stm32_init(&loader_config);
+    return ERR_OK;
+}
+
+/**
+ * \brief           USART data processing
+ */
+static void uart_reader(void const *arg) {
+    struct netif *netif = (struct netif *)arg;
+
+    size_t pos = 0;
+
+    log_info(ESPIF, "Reader thread running");
+
+    while (1) {
+        /* Wait for the event message from DMA or USART */
+        /* There is 100ms max wait time to ensure some small standalone message
+           does not get stuck in DMA buffer for too long. */
+        osMessageGet(uartBufferMbox_id, 100);
+
+        // TODO: Some kind of locking/safety guarantees on old_dma_pos
+
+        /* Read data */
+        uint32_t dma_bytes_left = __HAL_DMA_GET_COUNTER(huart6.hdmarx); // no. of bytes left for buffer full
+        pos = sizeof(dma_buffer_rx) - dma_bytes_left;
+        if (pos != old_dma_pos && espif_get_operating_mode() == ESPIF_RUNNING_MODE) {
+            if (pos > old_dma_pos) {
+                uart_input(&dma_buffer_rx[old_dma_pos], pos - old_dma_pos, netif);
+            } else {
+                uart_input(&dma_buffer_rx[old_dma_pos], sizeof(dma_buffer_rx) - old_dma_pos, netif);
+                if (pos > 0) {
+                    uart_input(&dma_buffer_rx[0], pos, netif);
+                }
+            }
+            old_dma_pos = pos;
+            if (old_dma_pos == sizeof(dma_buffer_rx)) {
+                old_dma_pos = 0;
+            }
+        }
+    }
+}
+
+static void process_mac(uint8_t len, uint8_t *data, struct netif *netif) {
+    log_info(ESPIF, "MAC len: %d, data: %02x%02x%02x%02x%02x%02x", len, data[0], data[1], data[2], data[3], data[4], data[5]);
+    xTaskNotify(init_task_handle, 0, eNoAction);
+    if (len != ETHARP_HWADDR_LEN) {
+        log_error(ESPIF, "Invalid MAC len: %d, ignoring", len);
+        return;
+    }
+    netif->hwaddr_len = len;
+    memcpy(netif->hwaddr, data, len);
```

Comment:
Well, this is a bit corrupted part of the ESP<>STM cooperation. First, this is called only from reader thread, no sync is needed. Ideally, ESP would send its MAC on reset while letting STM know it lives and allowing it to configure netif according to the MAC. Unfortunately the ESP cannot send MAC before it is configured to join AP and it can be configured this way exactly once. To cope with the limitation the ESP sends zero length MAC to show it lives and this is the reason the length is checked after init is unblocked. There is a hope that sane ESP firmware will not have this limitation.